cidc_api/utils/migrations.py

```diff
@@ -0,0 +1,159 @@
+
+            # Regenerate additional metadata from the migrated clinical trial
+            # metadata object.
+            df.additional_metadata = _get_uuid_info(
+                migration.result, artifact["upload_placeholder"]
+            )[1]
+
+            # If the GCS URI has changed, rename the blob
+            new_gcs_uri = artifact["object_url"]
+            if old_gcs_uri != new_gcs_uri:
+                renamer = PieceOfWork(
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, old_gcs_uri, new_gcs_uri
+                    ),
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, new_gcs_uri, old_gcs_uri
+                    ),
+                )
+                gcs_tasks.schedule(renamer)
+
+    # Migrate all assay upload successes
+    successful_assay_uploads: List[AssayUploads] = session.query(
+        AssayUploads
+    ).filter_by(status=AssayUploadStatus.MERGE_COMPLETED.value).with_for_update().all()
+    for upload in successful_assay_uploads:
+        migration = metadata_migration(upload.assay_patch)
+
+        # Update the metadata patch
+        upload.assay_patch = migration.result
+
+        # Update the GCS URIs of files that were part of this upload
+        old_file_map = upload.gcs_file_map
+        new_file_map = {}
+        for (
+            old_upload_uri,
+            old_target_uri,
+            artifact_uuid,
+        ) in upload.upload_uris_with_data_uris_with_uuids():
+            upload_timestamp = old_upload_uri[len(old_target_uri) + 1 :]
+            if old_target_uri in migration.file_updates:
+                new_target_uri = migration.file_updates[old_target_uri]["object_url"]
+                if old_gcs_uri != new_target_uri:
+                    new_upload_uri = "/".join([new_target_uri, upload_timestamp])
+                    renamer = PieceOfWork(
+                        lambda: rename_gcs_blob(
+                            GOOGLE_UPLOAD_BUCKET, old_upload_uri, new_upload_uri
+                        ),
+                        lambda: rename_gcs_blob(
+                            GOOGLE_UPLOAD_BUCKET, new_upload_uri, old_upload_uri
+                        ),
+                    )
+                    gcs_tasks.schedule(renamer)
+                new_file_map[new_upload_uri] = artifact_uuid
+
+        # Update the upload's file map to use new GCS URIs
+        upload.gcs_file_map = new_file_map
+
+    # Migrate all manifest records
+    manifest_uploads: List[ManifestUploads] = session.query(
+        ManifestUploads
+    ).with_for_update().all()
+    for upload in manifest_uploads:
+        migration = metadata_migration(upload.metadata_patch)
+
+        # Update the metadata patch
+        upload.metadata_patch = migration.result
+
+    # Attempt to make GCS updates
+    gcs_tasks.run_all()
+
+    # Update the database if everything else has succeeded
+    session.commit()
```

Comment:
so I feel like all those `with_for_update().all` might hold locks up until here, which seems like a lot of time / work in between.

Reply:
(we discussed offline and concluded that since only cloud functions would potentially get blocked by this lock, and cfns are background tasks, we're ok with this for now)